src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+            {
+                this.logger.LogDebug("Chained header '{0}' does not match checkpoint '{1}'.", chainedHeader, checkpoint.Hash);
+                this.logger.LogTrace("(-)[INVALID_HEADER_NOT_MATCHING_CHECKPOINT]");
+                throw new InvalidHeaderException();
+            }
+
+            ChainedHeader subchainTip = chainedHeader;
+            if (chainedHeader.Height == this.checkpoints.GetLastCheckpointHeight())
+                subchainTip = latestNewHeader;
+
+            ConnectedHeaders connectedHeaders = this.MarkBetterChainAsRequired(subchainTip);
+            this.MarkTrustedChainAsAssumedValid(chainedHeader);
+
+            this.logger.LogTrace("(-):{0}", connectedHeaders);
+            return connectedHeaders;
+        }
+
+        /// <summary>
+        /// Check whether a header is in one of the following states
+        /// <see cref="BlockDataAvailabilityState.BlockAvailable"/>, <see cref="BlockDataAvailabilityState.BlockRequired"/>.
+        /// </summary>
+        private bool HeaderWasNotRequested(ChainedHeader chainedHeader)
+        {
+            return !((chainedHeader.BlockDataAvailability == BlockDataAvailabilityState.BlockAvailable)
+                  || (chainedHeader.BlockDataAvailability == BlockDataAvailabilityState.BlockRequired));
+        }
+
+        /// <summary>
+        /// Check whether a header is in one of the following states
+        /// <see cref="ValidationState.AssumedValid"/>, <see cref="ValidationState.PartiallyValidated"/>, <see cref="ValidationState.FullyValidated"/>.
+        /// </summary>
+        private bool HeaderWasNotMarkedAsValidated(ChainedHeader chainedHeader)
+        {
+            return !((chainedHeader.BlockValidationState == ValidationState.AssumedValid)
+                  || (chainedHeader.BlockValidationState == ValidationState.PartiallyValidated)
+                  || (chainedHeader.BlockValidationState == ValidationState.FullyValidated));
+        }
+
+        private void RemoveUnclaimedBranch(ChainedHeader chainedHeader)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeader), chainedHeader);
+
+            ChainedHeader currentHeader = chainedHeader;
+            while (true)
+            {
+                bool headerHasNextHeader = currentHeader.Next.Count > 0;
+                if (headerHasNextHeader)
+                {
+                    this.logger.LogTrace("Header '{0}' is part of another branch.", currentHeader);
+                    break;
+                }
+
+                bool headerHasPeerClaim = this.peerIdsByTipHash.ContainsKey(chainedHeader.HashBlock);
+                if (headerHasPeerClaim)
+                {
+                    this.logger.LogTrace("Header '{0}' is claimed by a peer and wont be removed.", currentHeader);
+                    break;
+                }
+
+                this.logger.LogTrace("Header '{0}' was removed from the chain tree.", currentHeader);
```

Comment:
I started to dislike `chain tree`, I would just use `tree` as it makes better sense to me. Not only here.

Reply:
I would also put this log after the 2 lines that follow.